### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by sungam3r**
@Shane32  Please review this test for block string. Also I have a question about one case left:
изображение

Is such a situation possible? If so, could you give an example of a test that should be added? If no I suggest to remove that case.

### CODE REVIEW

**Comment by sungam3r** [comment]
Only one line uncovered
изображение

**Comment by Shane32** [comment]
I’ll try to review tomorrow or maybe tonight.

**Comment by sungam3r** [comment]
Also found this
изображение

I repeat, I never had a patience to understand all these transcoding 😞 .

**Comment by Shane32** [comment]
I really enjoy those types of coding puzzles. But I’ve been concentrating on the auto registering graph types

File: src/GraphQLParser/Exceptions/GraphQLParserException.cs [hunk 1]
```diff
@@ -102,10 +103,4 @@ private static bool HasReplacementCharacter(string str)
     }
 
     private static bool IsReplacementCharacter(char code) => code < 0x0020 && code != 0x0009 && code != 0x000A && code != 0x000D;
-
-    private static string GetUnicodeRepresentation(char code) => code switch
-    {
-        '\0' => "\\u0000",
```
Comment (sungam3r) [inline src/GraphQLParser/Exceptions/GraphQLParserException.cs:108]: I don't know why does this line here. `((int)'\0').ToString("D4")` gives `"\\u0000"` as well.

File: src/GraphQLParser/ParserContext.Parse.cs [hunk 2]
```diff
@@ -1328,22 +1328,20 @@ private GraphQLType ParseType()
...
+
+            nonNull.Type = type;
+            // move comment from wrapped type to wrapping type
+            nonNull.Comments = type.Comments;
+            type.Comments = null;
+            nonNull.Location = GetLocation(start);
+            type = nonNull;
         }
 
-        var nonNull = NodeHelper.CreateGraphQLNonNullType(_ignoreOptions); //TODO: deal with depth
```
Comment (sungam3r) [inline src/GraphQLParser/ParserContext.Parse.cs:1337]: TODO resolved, nothing to fix here, I just invert condition to align count of `IncreaseDepth` calls with count of `DecreaseDepth` calls.

**Comment by Shane32** [comment]
I spent some time looking at this.  I think what we need to decide is whether or not we want strings to always parse to exactly as input.

For instance, there is no way to encode any of the following text in a blockstring:
- `"\nTest"` - first line contains only whitespace
- "Test\n" - last line contains only whitespace
- `"Line 1\r\nLine 2"` - newlines (meaning `"\r\n", "\r", and "\n") coalesce to `\n` within a blockstring
- Strings containing control characters except certain ones

It is also noteworthy that some string sequences require a trailing blank line when formatted as a blockstring.  Specifically:
- Any string that ends with a quote must be encoded with a trailing newline in order to be properly parsed.
- If the text has uniform indentation

If we assume that the initial `"""` and trailing `"""` always is on its own line, with the trailing `"""` indented properly, then we can simplify checks a bit.  The rules now become something like this (to ensure that a string will parse exactly as input):
- Must not have beginning or trailing lines that are empty or contain only whitespace (tab or space)
- Cannot contain `"\r"` at all
- Cannot contain control characters unsupported by the GraphQL spec as SourceCharacter

I suggest we add an option to the printer to allow for some or all of the following:
- Allow newline conversion (`"\r\n"` and `"\r"` ->`"\n"`)
- Allow removing starting or ending blank lines

I'm not sure exactly what to call these options, and if we should have them separate or a single option.  Also I would assume that when we print an AST, only the description would be subject to such changes.  I would expect that printed string literals as part of a query, for instance, would always ensure that they can parse to exactly what was input.

**Comment by Shane32** [comment]
As for the line of code in question, the input should be `"Test\r\nLine 2\rLine 3\nLine 4"` and the output should be:

```
"""
Test
Line 2
Line 3
Line 4
"""
```

Unless I miss my guess, however, this would print something like this currently:

```
"""
Test
Line 2Line 3
Line 4
"""
```

It would "cover" the line of code you mentioned, but I would consider that behavior incorrect -- but there is no way to encode `"\r"` in a blockstring at all, as previously discussed.  So even if this is "fixed", the output still will not match the input, and would produce `"Test\nLine 2\nLine 3\nLine 4"` when decoded.

File: src/GraphQLParser/Exceptions/GraphQLParserException.cs (hunk 1 above)
Comment (Shane32) [reply src/GraphQLParser/Exceptions/GraphQLParserException.cs:108]: Probably just an optimization for null characters to prevent an allocation or something.

File: src/GraphQLParser.Tests/Visitors/SDLPrinterTests.cs [hunk 3]
```diff
@@ -580,6 +580,8 @@ type Query
     [InlineData(28, "Test \\n escaping", "Test \\\\n escaping", false)]
     [InlineData(29, "Test \\u1234 escaping", "Test \\\\u1234 escaping", false)]
     [InlineData(30, "Test \\ escaping", "Test \\\\ escaping", false)]
+    [InlineData(31, "t\"est\n  line2", "\nt\"est\nline2\n", true)] // BlockString with double quote inside
+    [InlineData(32, "t\\\"\"\"est\n  line2", "\nt\\\"\"\"est\nline2\n", true)] // BlockString with triple double quote inside
     public async Task SDLPrinter_Should_Print_BlockStrings(int number, string input, string expected, bool isBlockString)
```
Comment (Shane32) [inline src/GraphQLParser.Tests/Visitors/SDLPrinterTests.cs:585]: These new test conditions look good.

We cannot test for a string with a CR in it because all strings here are interpreted as a blockstring.  Since a blockstring cannot contain a CR (it would be parsed as a LF), it is impossible to test here what happens when the code tries to serialize a string containing a CR.  We would need to add it to another test.  We also cannot test for beginning or trailing whitespace and so on.  All that we can test for here is strings that can be encoded as blockstrings to begin with.

File: src/GraphQLParser/Visitors/PrintContextExtensions.cs [hunk 4]
```diff
@@ -21,7 +21,7 @@ public static ValueTask WriteAsync<TContext>(this TContext context, ROM value)
         var task =
 #if NETSTANDARD2_0
         // no cancellationToken support on netstandard2.0
-        context.Writer.WriteAsync(value.ToString());
+        context.Writer.WriteAsync(value.ToString()); //ISSUE: allocation - either WriteAsync(value.ToString()) or Write(char value) in a loop
```
Comment (Shane32) [inline src/GraphQLParser/Visitors/PrintContextExtensions.cs:24]: Calling `Write(char value)` in a loop would be a performance killer.  Much better to have an allocation if needed.

**Review by Shane32 — APPROVED** [review]

File: src/GraphQLParser/Visitors/PrintContextExtensions.cs (hunk 4 above)
Comment (sungam3r) [reply src/GraphQLParser/Visitors/PrintContextExtensions.cs:24]: ok

**Comment by sungam3r** [comment]
@Shane32  Thanks for test example. I added it to get full coverare (correct or not). Unfortunately, I understood little from your explanation. I confess that the transcoding is not my strength. Let's add the necessary options and change the output if you consider it necessary.

File: src/GraphQLParser/Exceptions/GraphQLParserException.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQLParser/Exceptions/GraphQLParserException.cs:108]: Looks strange among all other string allocations above.

**Comment by sungam3r** [comment]
`SDLPrinter` 100% covered
